### DISCUSSION TITLE
Help creating driver for Itron Equascan Cost Allocator

### DISCUSSION THREAD

**Discussion by miguelrojorossi** [wmbusmeters/wmbusmeters#1401] [Q&A]
Hi,

I have several Itron Equascan eHCA V2 Cost Allocators at home and I am trying to decode the telegrams, but there is no matching driver.

I have managed to scan some telegrams:

wmbusmeters --logtelegrams --format=json /dev/ttyUSB0:im871a[00104738]:c1 Heat auto 23203225 NOKEY
Started config im871a[00104738] on /dev/ttyUSB0 listening on c1
(meter) Heat: meter detection could not find driver for id: 23203225 mfct: (ITW) ITRON Water, France (0x2697) type: Heat Cost Allocator (0x08) ver: 0x00
(meter) please consider opening an issue at https://github.com/wmbusmeters/wmbusmeters/
(meter) to add support for this unknown mfct,media,version combination
telegram=|2C4497262532202300087ACC0000A0_0B6E000000066D03070E183A004B6E270100426C1E36067F2C008F290706|+150
{"media":"heat cost allocation","meter":"auto","name":"Heat","id":"23203225","timestamp":"2024-10-24T13:05:00Z","device":"im871a[00104738]","rssi_dbm":-60}
telegram=|2C4497262532202300087ACD0000A0_0B6E000000066D030A0E183A004B6E270100426C1E36067F2C008F290706|+329
{"media":"heat cost allocation","meter":"auto","name":"Heat","id":"23203225","timestamp":"2024-10-24T13:07:59Z","device":"im871a[00104738]","rssi_dbm":-60}
telegram=|2C4497262532202300087ACE0000A0_0B6E000000066D030D0E183A004B6E270100426C1E36067F2C008F290706|+509
{"media":"heat cost allocation","meter":"auto","name":"Heat","id":"23203225","timestamp":"2024-10-24T13:10:59Z","device":"im871a[00104738]","rssi_dbm":-60}
telegram=|2C4497262532202300087ACF0000A0_0B6E000000066D03100E183A004B6E270100426C1E36067F2C008F290706|+689
{"media":"heat cost allocation","meter":"auto","name":"Heat","id":"23203225","timestamp":"2024-10-24T13:13:59Z","device":"im871a[00104738]","rssi_dbm":-59}
telegram=|2C4497262532202300087AD00000A0_0B6E000000066D03130E183A004B6E270100426C1E36067F2C008F290706|+869
{"media":"heat cost allocation","meter":"auto","name":"Heat","id":"23203225","timestamp":"2024-10-24T13:16:59Z","device":"im871a[00104738]","rssi_dbm":-60}
telegram=|2C4497262532202300087AD10000A0_0B6E000000066D03160E183A004B6E270100426C1E36067F2C008F290706|+1049
{"media":"heat cost allocation","meter":"auto","name":"Heat","id":"23203225","timestamp":"2024-10-24T13:19:59Z","device":"im871a[00104738]","rssi_dbm":-52}

As well as analyzed some of them, and can read most of the telegram:

wmbusmeters --analyze=auto 2C4497262532202300087AB00000A0_0B6E000000066D032B0C183A004B6E270100426C1E36067F2C008F290706
Auto driver    : not found!
Similar driver : qcaloric 14/20
Using driver   : qcaloric 14/20
000   : 2c length (44 bytes)
001   : 44 dll-c (from meter SND_NR)
002   : 9726 dll-mfct (ITW)
004   : 25322023 dll-id (23203225)
008   : 00 dll-version
009   : 08 dll-type (Heat Cost Allocator)
010   : 7a tpl-ci-field (EN 13757-3 Application Layer (short tplh))
011   : b0 tpl-acc-field
012   : 00 tpl-sts-field (OK)(OK)
013   : 00a0 tpl-cfg a000 (bidirectional synchronous )
015   : 0B dif (6 digit BCD Instantaneous value)
016   : 6E vif (Units for H.C.A.)
017 C!: 000000 ("current_consumption_hca":0)
020   : 06 dif (48 Bit Integer/Binary Instantaneous value)
021   : 6D vif (Date and time type)
022 C!: 032B0C183A00 ("device_date_time":"2024-10-24 12:43:03")
028   : 4B dif (6 digit BCD Instantaneous value storagenr=1)
029   : 6E vif (Units for H.C.A.)
030 C!: 270100 ("consumption_at_set_date_hca":127) ("consumption_at_set_date_1_hca":127)
033   : 42 dif (16 Bit Integer/Binary Instantaneous value storagenr=1)
034   : 6C vif (Date type G)
035 C!: 1E36 ("set_date":"2024-06-30") ("set_date_1":"2024-06-30")
037   : 06 dif (48 Bit Integer/Binary Instantaneous value)
038   : 7F vif (Manufacturer specific)
039 C?: 2C008F290706

{
    "media":"heat cost allocation",
    "meter":"qcaloric",
    "name":"",
    "id":"23203225",
    "consumption_at_set_date_hca":127,
    "consumption_at_set_date_1_hca":127,
    "current_consumption_hca":0,
    "device_date_time":"2024-10-24 12:43:03",
    "set_date":"2024-06-30",
    "set_date_1":"2024-06-30",
    "status":"OK",
    "timestamp":"2024-10-24T12:15:06Z"
}

Where can I find how to create a driver? I cannot find it on the wiki. Or can someone help me create one?

Thanks!!

**Comment by miguelrojorossi** [comment]
I have also tried with the itronheat driver with no luck:

[2024-10-24_22:49:59] (meter) Repartidores: meter detection did not match the selected driver itronheat! correct driver is: unknown!
(meter) Not printing this warning again for id: 23203225 mfct: (ITW) ITRON Water, France (0x2697) type: Heat Cost Allocator (0x08) ver: 0x00
[2024-10-24_22:49:59] (meter) please consider opening an issue at https://github.com/wmbusmeters/wmbusmeters/

The actual output from that driver is missing lots of info compared to using the qcaloric one:

{"media":"heat cost allocation","meter":"itronheat","name":"Repartidores","id":"23203225","last_year_date":"2024-06-30","meter_datetime":"2024-10-24 21:52:03","status":"OK","timestamp":"2024-10-24T20:49:59Z","device":"im871a[00104738]","rssi_dbm":-54}

**Comment by miguelrojorossi** [comment]
I have opened a device support issue

**Comment by BIBOLV** [comment]
Hi,
Did you confirm that readings that qcaloric driver provided are correct? If yes, I would suggest to use that driver for future, as there are only small part of data at the end that was not parsed by driver and deemed "Manufacturer specific".
